update-v1.c

```diff
@@ -0,0 +1,376 @@
+		fprintf(stderr, "This update is for a %04X, not a %04X.\n", ftr.model, board->modelnum);
+		return 1;
+	}
+
+	/* gcc warns this pointer has alignment issues in packed structure. */
+	bin_size = (uint32_t)ftr.bin_size;
+
+	/* Check file is 128-byte aligned */
+	if (bin_size & 0x7F)
+		error(1, 0, "Binary file must be 128-byte aligned!");
+
+	fflush(stdout);
+
+	/*
+	 * Let the messages print out. Some of the flash operations will
+	 * cause the micro to drop some chars if they output while we touch 
+	 * flash 
+	 */
+	usleep(1000 * 10);
+
+	/* Write magic key and length/location information */
+	if (spokestream16(i2cfd, board->i2c_chip, SUPER_FL_MAGIC_KEY0, (uint16_t *)&magic_key, 4) < 0)
+		error(1, errno, "Failed to write magic key");
+
+	if (spokestream16(i2cfd, board->i2c_chip, SUPER_FL_SZ0, (uint16_t *)&bin_size, 4) < 0)
+		error(1, errno, "Failed to bin length");
+
+retry:
+	lseek(binfd, 0, SEEK_SET);
+
+	if (retries == 0) {
+		fprintf(stderr, "Failed to update microcontroller, contact support\n");
+		return 1;
+	}
+	retries--;
+
+	/* If flash is already opened from a previous action, close it to reset the flash state. */
+	status = speek16(i2cfd, board->i2c_chip, SUPER_FL_FLASH_STS) & 0xff;
+	if (status != STATUS_CLOSED) {
+		spoke16(i2cfd, board->i2c_chip, SUPER_FL_FLASH_CMD, SUPER_CLOSE_FLASH);
+		status = speek16(i2cfd, board->i2c_chip, SUPER_FL_FLASH_STS) & 0xff;
+		if (status != STATUS_CLOSED) {
+			fprintf(stderr, "Couldn't re-close flash!\n");
+			goto retry;
+		}
+	}
+
+	/* Poll until flash is opened. This also has to check/erase flash
+	 * which happens while interrupts are disabled for flash safety. Because
+	 * interrupts are disabled, I2C transactions get stalled, and can
+	 * generate errors. Wait a long timeout before trying to talk to the
+	 * uC again.
+	 */
+	spoke16(i2cfd, board->i2c_chip, SUPER_FL_FLASH_CMD, SUPER_OPEN_FLASH);
+	sleep(1);
+	status = speek16(i2cfd, board->i2c_chip, SUPER_FL_FLASH_STS) & 0xff;
+	if (status != STATUS_READY) {
+		fprintf(stderr, "Failed to open flash!\n");
+		if (status != STATUS_CLOSED)
+			flash_print_error(status);
+		goto retry;
+	}
+
+	/* Write BIN to MCU via I2C */
+	for (i = bin_size; i; i -= 128) {
+		printf("\r%d/%d", bin_size - i, bin_size);
+		fflush(stdout);
+		ret = read(binfd, buf, 128);
+		if (ret < 0) {
+			fprintf(stderr, "Error: %s, Error reading from BIN @ %d", strerror(errno), bin_size - i);
+			goto retry;
+		} else if (ret < 128) {
+			fprintf(stderr, "Error: short read from  bin, got %d instead of 128\n", ret);
+			goto retry;
+		} else {
+			crc = (uint16_t)crc8((uint8_t *)buf, 128);
+
+			/*
+			 * Prefer streaming interface, but fall back to individual pokes if
+			 * larger writes are failing (might be interrupted?)
+			 */
+			if (retries > 5)
+				spokestream16(i2cfd, board->i2c_chip, SUPER_FL_BLOCK_DATA, buf, 128);
+			else
+				for (int x = 0; x < 64; x++)
+					spoke16(i2cfd, board->i2c_chip, SUPER_FL_BLOCK_DATA + x, buf[x]);
```

Comment:
I don't know how to feel about this if clang-format called this good :rofl:

Reply:
I removed retries, so this is resolved as a part of this :P